Handle malformed tokens with jiffy 1.x

jaydoane opened the pull request:
Recent changes in how `jiffy:decode/1` handles malformed JSON has caused
`jwtf:decode/3` to fail to properly return a bad request 400 response
for some malformed tokens.

First, this changes the name of the function to `decode_b64url_json/1`,
indicating that it decodes something that has been first been JSON
encoded, and then base64 encoded.

More substantially, it wraps both the base64url and jiffy decoding in a
try/catch block, since both can throw errors, while the former can also
return an error tuple. Tests have been added to ensure all code paths
are covered.

jaydoane committed:
Enable code coverage

rnewson commented on src/jwtf.erl:
```diff
@@ -102,10 +102,10 @@ verification_algorithm(Alg) ->
 
 validate(Header0, Payload0, Signature, Checks, KS) ->
-    Header1 = props(decode_json(Header0)),
+    Header1 = props(decode_b64_json(Header0)),
```
say `b64url` here as `b64` is a slightly different encoding.

rnewson replied:
(and in the commit text, etc)

jaydoane replied:
done

rnewson commented on src/jwtf.erl:
```diff
@@ -269,14 +269,20 @@ split(EncodedToken) ->
+        case b64url:decode(B64Encoded) of
+            {error, Reason} ->
+                throw({bad_request, Reason});
+            JsonEncoded ->
```
b64url:decode is quite naughty for having a `binary() | {error, term()}` result, but it is what it is...

jaydoane replied:
yeah, it's super annoying that it can both return an `{error, term()}` tuple and also throw exceptions

rnewson reviewed (approved):
good improvement.

jaydoane committed:
Handle malformed tokens with jiffy 1.x

Recent changes in how `jiffy:decode/1` handles malformed JSON has caused
`jwtf:decode/3` to fail to properly return a bad request 400 response
for some malformed tokens.

First, this changes the name of the function to `decode_b64url_json/1`,
indicating that it decodes something that has been first been JSON
encoded, and then base64url encoded.

More substantially, it wraps both the base64url and jiffy decoding in a
try/catch block, since both can throw errors, while the former can also
return an error tuple. Tests have been added to ensure all code paths
are covered.
